### PR TITLE
[Tests] Use the reduce Method to Analyze Data

### LINKED ISSUE #18104: Bug in freecodecamp functional programming section
#### Describe your problem and - if possible - how to reproduce it

 I sat over this for a bit trying to figure out why the average wasn’t coming out to what the exercise required. So i messed around a bit. Only after the code passed realized it wanted the average of all the movies by a certain director, and I fixed it. however
this monstrosity did pass. even though it shouldn’t, because it doesnt do what the exercise requires… Just saying…

```js
var averageRating;

var arr=[];
var rating = watchList.map(function(obj){ return arr.push(parseFloat(obj.imdbRating));
})
console.log(arr);
var averageRating= Math.floor((arr.reduce(function(a,b){
return a+b;
},)/arr.length+0.1555)*1000)/1000;
```

#### Add a Link to the page with the problem
https://learn.freecodecamp.org/javascript-algorithms-and-data-structures/functional-programming/use-the-reduce-method-to-analyze-data

#### Tell us about your browser and operating system
Chrome, Windows 10
* Browser Name: 
* Browser Version: 
* Operating System:

### PR DESCRIPTION
**Pull request by ezioda004**
Added a function in seed, changed the description accordingly and added additional test. Successfully tested locally.

- [x] I have read [freeCodeCamp's contribution guidelines](https://github.com/freeCodeCamp/freeCodeCamp/blob/master/CONTRIBUTING.md).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `master` branch of freeCodeCamp.
- [x] None of my changes are plagiarized from another source without proper attribution.
- [x] My article does not contain shortened URLs or affiliate links.

If your pull request closes a GitHub issue, replace the XXXXX below with the issue number.

Closes #18104

### CODE REVIEW

**Comment by ezioda004** [comment]
@ValeraS Can you check this PR? I've successfully tested locally. Thanks.

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/functional-programming/use-the-reduce-method-to-analyze-data.english.md [hunk 1]
```diff
@@ -155,13 +157,18 @@ var watchList = [
                 }
 ];
 
-// Add your code below this line
+function getRating(){
```
Comment (ValeraS) [inline curriculum/challenges/english/02-javascript-algorithms-and-data-structures/functional-programming/use-the-reduce-method-to-analyze-data.english.md]: IMO, it should be the next:

```js
function getRating(watchList) {
```

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/functional-programming/use-the-reduce-method-to-analyze-data.english.md [hunk 2]
```diff
@@ -155,13 +157,18 @@ var watchList = [
...
+  // Add your code below this line
+  var averageRating;
+
+
+  console.log(averageRating);
```
Comment (ValeraS) [inline curriculum/challenges/english/02-javascript-algorithms-and-data-structures/functional-programming/use-the-reduce-method-to-analyze-data.english.md]: remove or place bellow `console.log(getRating(watchList));`

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/functional-programming/use-the-reduce-method-to-analyze-data.english.md [hunk 3]
```diff
@@ -155,13 +157,18 @@ var watchList = [
...
+  return averageRating;
+}
 
-var averageRating;
 
 // Add your code above this line
```
Comment (ValeraS) [inline curriculum/challenges/english/02-javascript-algorithms-and-data-structures/functional-programming/use-the-reduce-method-to-analyze-data.english.md]: place it above `return`

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/functional-programming/use-the-reduce-method-to-analyze-data.english.md [hunk 4]
```diff
@@ -174,6 +181,131 @@ console.log(averageRating);
...
+                   "imdbRating": "7.9",
+                   "imdbVotes": "876,575",
+                   "imdbID": "tt0499549",
+                   "Type": "movie",
+                   "Response": "True"
+                }
+];
+
+function getRating(){
+  var averageRating = watchList.reduce((a, c) => {
```
Comment (ValeraS) [inline curriculum/challenges/english/02-javascript-algorithms-and-data-structures/functional-programming/use-the-reduce-method-to-analyze-data.english.md]: IMO, it is better to use `filter` with` map` and after `reduce`. The use of `filter` and` map` is suggested in the instructions.
Something like:
```js
const ratings = watchList.filter(...).map(...);
averageRating = ratings.reduce(...)/ratings.length;
```

**Review by ValeraS — CHANGES_REQUESTED** [review]

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/functional-programming/use-the-reduce-method-to-analyze-data.english.md (hunk 4 above)
Comment (ValeraS) [reply curriculum/challenges/english/02-javascript-algorithms-and-data-structures/functional-programming/use-the-reduce-method-to-analyze-data.english.md]: Of course, it's a correct solution but, I think we should provide solutions as an example of "good" code. Next who will see this code will not need to understand what `a`, `c`, `r` mean and why you assign `averageRating` in `return`.

Comment (ezioda004) [reply curriculum/challenges/english/02-javascript-algorithms-and-data-structures/functional-programming/use-the-reduce-method-to-analyze-data.english.md]: Agreed. I thought the solution in curriculum is for testing only? 
The reference solutions are provided in the `guide` articles.

Ofcourse thats no excuse for `a`, `c`, `r`.

Fixed.

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/functional-programming/use-the-reduce-method-to-analyze-data.english.md [hunk 5]
```diff
@@ -26,10 +26,12 @@ tests:
...
   - text: Your code should use the <code>reduce</code> method.
     testString: assert(code.match(/\.reduce/g), 'Your code should use the <code>reduce</code> method.');
-  - text: The <code>averageRating</code> should equal 8.675.
-    testString: assert(averageRating == 8.675, 'The <code>averageRating</code> should equal 8.675.');
+  - text: The <code>getRating(watchList)</code> should equal 8.675.
+    testString: assert(getRating(watchList) === 8.675, 'The <code>getRating(watchList)</code> should equal 8.675.');
   - text: Your code should not use a <code>for</code> loop.
     testString: assert(!code.match(/for\s*?\(.*\)/g), 'Your code should not use a <code>for</code> loop.');
+  - text: Your code should return correct output after modifying the <code>watchList</code> object.
+    testString: assert((() => { watchList.splice(1, 2);  return getRating(watchList) === 8.55;})(), 'Your code should return correct output after modifying the <code>watchList</code> object');
```
Comment (ValeraS) [inline curriculum/challenges/english/02-javascript-algorithms-and-data-structures/functional-programming/use-the-reduce-method-to-analyze-data.english.md]: Let's simplify it:
```js
assert(getRating(watchList.slice(...)) === ..., '')
```

**Review by ValeraS — CHANGES_REQUESTED** [review]

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/functional-programming/use-the-reduce-method-to-analyze-data.english.md (hunk 5 above)
Comment (ValeraS) [reply curriculum/challenges/english/02-javascript-algorithms-and-data-structures/functional-programming/use-the-reduce-method-to-analyze-data.english.md]: You could use a `filter`.

Comment (ezioda004) [reply curriculum/challenges/english/02-javascript-algorithms-and-data-structures/functional-programming/use-the-reduce-method-to-analyze-data.english.md]: Like this?
```
getRating(watchList.filter((obj, i) => i === 0 || i > 2)) === 8.55 ?
```

Comment (ezioda004) [reply curriculum/challenges/english/02-javascript-algorithms-and-data-structures/functional-programming/use-the-reduce-method-to-analyze-data.english.md]: Done.

**Review by ValeraS — APPROVED** [review]

**Pull request merged**